fedbiomed/node/requests/_protocol_manager.py

```diff
@@ -0,0 +1,231 @@
+                            and not self._active_tasks[task_name]['finally']:
+                        # A task is expired: cancel it
+                        # Entry from the table will be cleaned by the `add_done_callback()`
+                        # after completing task `final()`
+                        self._active_tasks[task_name]['task'].cancel()
+
+                        # we cannot just rely on `task.cancel()` as it may re-cancel a task in its `final()` clause
+                        # if it does not complete it before next loop of this function
+                        # Basically, we want to issue *once* a cancel() to a task
+                        # and then trust it to properly complete
+                        self._active_tasks[task_name]['finally'] = True
+
+    async def _run_async(self) -> None:
+        """Main async function for the protocol manager background thread."""
+
+        self._loop = asyncio.get_running_loop()
+
+        # task for cleaning the active tasks
+        self._task_clean_active_tasks = asyncio.create_task(self._clean_active_tasks())
+
+        while True:
+            msg = await self._queue.get()
+            self._queue.task_done()
+
+            async with self._active_tasks_lock:
+                # TODO: add test for maximum number of tasks ?
+
+                # timeout for tasks in handled via `_clean_active_tasks` task
+                task_msg = asyncio.create_task(self._overlay_message_process(msg))
+                self._active_tasks[task_msg.get_name()] = {
+                    'start_time': time.time(),
+                    'task': task_msg,
+                    'finally': False,
+                }
+                #
+                task_msg.add_done_callback(self._remove_finished_task)
+
+    async def _submit(self, msg) -> None:
+        """Submits a received message to the protocol manager for processing.
+
+        Args:
+            msg: received message
+        """
+        try:
+            await self._queue.put_nowait(msg)
+        except asyncio.QueueFull as e:
+            logger.error(
+                f"{ErrorNumbers.FB324}: Failed submitting message to protocol manager. Discard message. "
+                f"Exception: {type(e).__name__}. Error message: {e}")
+            # don't raise exception
+
+    async def _overlay_message_process(self, overlay_msg: dict) -> None:
+        """Main function for a task processing a received message.
+
+        Args:
+            overlay_msg: received outer message
+        """
+
+        try:
+            try:
+                if overlay_msg['dest_node_id'] != environ['NODE_ID']:
+                    logger.error(
+                        f"{ErrorNumbers.FB324}: Node {environ['NODE_ID']} received an overlay message "
+                        f"sent to {overlay_msg['dest_node_id']}. Maybe malicious activity. Ignore message."
+                    )
+                    return
```

Comment:
Thank you @srcansiz 
Regarding which criterion is it "too late" ?

- from security point of view: IMO it is OK to treat it there because we did not yet do any "sensitive" processing of the message, eg using the payload of the message, acknowledging its origin. The real security stuff comes from `format_incoming_message()` where we later add signing/encryption.

- form an application logic point of view: this is to group here the code that has knowledge about the content of an overlay message. Previous checks know about the fields and types (`Message) or just check the command (`on_message`), so it would seem to me strange to add there some code with "knowledge" of the overlay message content if we can avoid it.

Another option would be to move this check in the `format_incoming_overlay` (just after) as it makes sense to consider it part of the incoming overlay message process ?